### PR TITLE
fix #17993 - Deprecated callable strings

### LINKED ISSUE #17993: PropertyAccessDecorator should not accept callable strings
Right now, PropertyAccessDecorator [treats callable strings as callables, not as strings](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php#L87). Consequently the following code does not behave as expected:

~~~php
$builder->add('deadline', ChoiceType::class, [
    'choices' => $dateRanges, // DateRange objects
    'choice_label' => 'end',
]);
~~~

This code results in a fatal error. The `end()` function exists (is callable), hence the string is not treated as property path. Instead, `end()` is called with invalid arguments. IMO there are very few cases where this would be the expected behavior, and even if, it could be solved by passing a closure:

~~~php
$builder->add('deadline', ChoiceType::class, [
    'choices' => $dateRanges, // arrays of dates
    'choice_label' => function (array $dateRange) {
        return end($dateRange);
    },
]);
~~~

Treating callable strings as callables should be deprecated. I.e., all methods of PropertyAccessDecorator should trigger deprecation errors `if (is_string($value) && is_callable($value))`.

### PR DESCRIPTION
**Pull request by Simperfit**
| Q | A |
| --- | --- |
| Branch | master |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | no |
| Fixed tickets | #17993 |
| License | MIT |
| Doc PR |  |

Is this ok ?
- [x] Rebase when  #18057 is merged

### CODE REVIEW

**Comment by Simperfit** [comment]
Thank you, I didn't understand the issue correctly.

I've changed my fix.

File: src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php [hunk 1]
```diff
@@ -163,6 +175,8 @@ public function createView(ChoiceListInterface $list, $preferredChoices = null,
 
         if (is_string($preferredChoices) && !is_callable($preferredChoices)) {
             $preferredChoices = new PropertyPath($preferredChoices);
+        } elseif(is_string($preferredChoices) && is_callable($preferredChoices)) {
```
Comment (Tobion) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: missing space after elseif

File: src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php [hunk 2]
```diff
@@ -111,12 +115,16 @@ public function createListFromChoices($choices, $value = null)
...
      *
+     * @deprecated We should not treat callable strings as callable
+     *
      * @return ChoiceListInterface The choice list
      */
     public function createListFromLoader(ChoiceLoaderInterface $loader, $value = null)
     {
         if (is_string($value) && !is_callable($value)) {
             $value = new PropertyPath($value);
+        } elseif(is_string($value) && is_callable($value)) {
```
Comment (Tobion) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: missing space after elseif

File: src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php [hunk 3]
```diff
@@ -80,12 +80,16 @@ public function getDecoratedFactory()
...
      *
+     * @deprecated We should not treat callable strings as callable
+     *
      * @return ChoiceListInterface The choice list
      */
     public function createListFromChoices($choices, $value = null)
     {
         if (is_string($value) && !is_callable($value)) {
             $value = new PropertyPath($value);
+        } elseif(is_string($value) && is_callable($value)) {
```
Comment (Tobion) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: missing space after elseif

File: src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php [hunk 4]
```diff
@@ -80,12 +80,16 @@ public function getDecoratedFactory()
      * @param null|callable|string|PropertyPath $value   The callable or path for
      *                                                   generating the choice values
      *
+     * @deprecated We should not treat callable strings as callable
```
Comment (Tobion) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: this should not be added as the whole method will be marked as deprecated

File: src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php (hunk 3 above)
Comment (Simperfit) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: I've changed it all

File: src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php (hunk 4 above)
Comment (Simperfit) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: Thank you,

In this case, I don't have to add any "@deprecated" at all ?

**Comment by HeahDude** [comment]
Not on these methods for sure.

File: src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php [hunk 5]
```diff
@@ -86,6 +86,8 @@ public function createListFromChoices($choices, $value = null)
     {
         if (is_string($value) && !is_callable($value)) {
             $value = new PropertyPath($value);
+        } elseif (is_string($value) && is_callable($value)) {
+            @trigger_error('createListFromChoices() treats callable strings as callable and is deprecated since version 3.0.', E_USER_DEPRECATED);
```
Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: since version 3.1 (for now)

**Comment by HeahDude** [comment]
You could perhaps wrap the callable string in a closure inside those `elseif` to prevent the fatal error.
```php
elseif (is_string($label) && is_callable($label)) {
    @trigger_error('createView() - $label - treats callable strings as callable and is deprecated since version 3.0.', E_USER_DEPRECATED);

    $value = function ($choice) use ($label) {
        return $label($choice);
    };
}

**Comment by Simperfit** [comment]
I have no errors doing the tests except of one in PasswordEncoder.

Thanks @HeahDude for the help

File: src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php [hunk 6]
```diff
@@ -86,6 +86,11 @@ public function createListFromChoices($choices, $value = null)
     {
         if (is_string($value) && !is_callable($value)) {
             $value = new PropertyPath($value);
+        } elseif (is_string($value) && is_callable($value)) {
+            @trigger_error('createListFromChoices() treats callable strings as callable and is deprecated since version 3.1.', E_USER_DEPRECATED);
```
Comment (iltar) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: If it's deprecated, what's the recommended thing to do instead? What will happen in 4.0? Please add this to the deprecation messages (and upgrade guide)

Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: @iltar is right. You should change each message to something like :
```php
'Treating strings as callable is deprecated since version 3.1 and will throw an error in 4.0. You should use a "\Closure" instead.'
```

**Comment by HeahDude** [comment]
>I have no errors doing the tests except of one in PasswordEncoder.

What error are you talking about ?

**Comment by Simperfit** [comment]
Nevermind, no error anymore

**Comment by fabpot** [comment]
@Simperfit What about adding some unit tests?

**Comment by Simperfit** [comment]
@fabpot I added one, will add one for each method

**Comment by xabbuh** [comment]
Looks like this needs another rebase. It contains some unrelated changes.

Status: Needs work

**Comment by Simperfit** [comment]
@xabbuh Sorry for the unrelated changes, I've missed my rebase.

File: src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php [hunk 7]
```diff
@@ -86,6 +86,14 @@ public function createListFromChoices($choices, $value = null)
...
         if (is_string($value) && !is_callable($value)) {
             $value = new PropertyPath($value);
+        } elseif (is_string($value) && is_callable($value)) {
+            @trigger_error('Treating strings as callable is deprecated since version 3.1 and will throw an error in 4.0. You should use a "\Closure" instead.', E_USER_DEPRECATED);
+            
+            $value = function (array $choices) use ($value) {
+                return $value($choices);
+            };
+
+            return $value($choices);
```
Comment (xabbuh) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: Do we need to do all the stuff beside triggering the deprecation? Shouldn't the following code already handle the callback properly?

Comment (Simperfit) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: I've followed the instruction of someone since I don't wanted to make things bad,
But you're right, adding the test without any code is the right thing to do.

Thanks

**Comment by Simperfit** [comment]
@fabpot I've added one test for each method, is it ok like that ? Thank you.

**Comment by HeahDude** [comment]
>Do we need to do all the stuff beside triggering the deprecation? Shouldn't the following code already handle the callback properly?

@xabbuh not when it is a callable string, see #17993.

Actually that's fine for `$value` (aka [`choice_value`](https://github.com/Simperfit/symfony/blob/master/src/Symfony/Component/Form/ChoiceList/ArrayChoiceList.php#L204)) as it gets the choice as only argument.

But all the others handled in `createView()` method ([`choice_name`](https://github.com/Simperfit/symfony/blob/master/src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php#L121), [`choice_label`](https://github.com/Simperfit/symfony/blob/master/src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php#L130), [`choice_attr`](https://github.com/Simperfit/symfony/blob/master/src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php#L140), [`preferred_choices`](https://github.com/Simperfit/symfony/blob/master/src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php#L144), [`group_by`](https://github.com/Simperfit/symfony/blob/master/src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php#L203)) would throw a fatal error because they get passed three arguments instead of one.

Consider @webmozart example:
```php
$builder->add('deadline', ChoiceType::class, [
    'choices' => $dateRanges, // DateRange objects
    'choice_label' => 'end',
    // One would expect $choice->getEnd() to be called
    // but this will never happen
    // because PropertyAccessDecorator treats it as a callable
    // and not as a string so it does not transform it to a closure.
]);
```
Instead `end($choice)` is called, fine for value but fatal for the others.

So I suggest to say "Treating callable strings as callable is deprecated since version 3.1 and PropertyAccessDecorator will treat them as strings in 4.0. You should use a "\Closure" instead.".

To avoid the fatal error for now, we should wrap the callable string in a closure:
```php
$label = function ($choice) use ($label) {
    return $label($choice);
}
```
So we get `call_user_func($label, $choice, $key, $value)` working until we can refactor in 4.0:
```php
if (is_string($value) && !is_callable($value)) {
    $value = new PropertyPath($value);
} elseif (is_string($value) && is_callable($value)) {
   // Trigger, handle...
}

// To simply

if (is_string($value)) {
    $value = new PropertyPath($value);
}
```

File: src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php [hunk 8]
```diff
@@ -86,6 +86,8 @@ public function createListFromChoices($choices, $value = null)
     {
         if (is_string($value) && !is_callable($value)) {
             $value = new PropertyPath($value);
+        } elseif (is_string($value) && is_callable($value)) {
+            @trigger_error('Treating strings as callable is deprecated since version 3.1 and will throw an error in 4.0. You should use a "\Closure" instead.', E_USER_DEPRECATED);
```
Comment (webmozart) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: "Passing callable strings is deprecated..."

File: src/Symfony/Component/Form/Tests/ChoiceList/Factory/PropertyAccessDecoratorTest.php [hunk 9]
```diff
@@ -63,6 +63,23 @@ public function testCreateFromChoicesPropertyPathInstance()
...
+     */
+    public function testCreateFromChoicesPropertyPathWithCallableString()
+    {
+        $choices = array(array('property' => 'value'));
+
+        $this->decoratedFactory->expects($this->once())
+            ->method('createListFromChoices')
+            ->with($choices, 'end')
+            ->will($this->returnCallback(function ($choices, $callback) {
+                return array_map($callback, $choices);
```
Comment (webmozart) [inline src/Symfony/Component/Form/Tests/ChoiceList/Factory/PropertyAccessDecoratorTest.php]: You can simplify this test by simply returning a string `'RESULT'` and asserting whether the return value matches that string. We don't actually need to call `array_map()`, that only complicates understanding the test.

**Comment by webmozart** [comment]
Thanks for working on this @Simperfit! :)

**Comment by xabbuh** [comment]
@HeahDude But if we solve it this way it means that we make a feature working while deprecating it at the same moment. If we think that we should wrap strings that are callables in a closure for safety, we should fix that issue in all affected branches (I guess `2.7` is the lowest affected branch, right?).

**Comment by HeahDude** [comment]
@xabbuh yes and no. The new feature would be:
```php
if (is_string($value) && !is_callable($value)) {
    $value = new PropertyPath($value);
} elseif (is_string($value) && is_callable($value)) {
    // Trigger
    $value = new PropertyPath($value);
}
```

But your right my actual patch should go on 2.7, so I send a PR and this one will need a rebase though.

File: src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php (hunk 8 above)
Comment (xabbuh) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: Should this really throw an exception in Symfony 4.0? I mean should we make the distinction of callable and not callable strings? Why not treat all strings as property paths then?

Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: @xabbuh see my comment above:
>So I suggest to say "Treating callable strings as callable is deprecated since version 3.1 and PropertyAccessDecorator will treat them as strings in 4.0. You should use a "\Closure" instead.".

Comment (xabbuh) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: @HeahDude :+1:

Comment (Simperfit) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: @xabbuh @HeahDude @webmozart I've modified it all

**Comment by Simperfit** [comment]
@webmozart thanks to you for helping me :)

@HeahDude I was looking into the PR you made but I can't make the test working, using both closure or PropertyPath, do you know what I'm missing ? Thank you for helping me from the start.

**Comment by HeahDude** [comment]
@Simperfit You don't need to worry about that anymore.

If validated, once #18057 is merged in master, you will have to rebase, then your commit will only need to add deprecation notices by triggering errors and marking tests in legacy group.

You should mark this PR as "WIP" for now and add a todo to watch #18057, so someone could check it and ping you when the rebase is needed.

Cheers

File: src/Symfony/Component/Form/Tests/ChoiceList/Factory/PropertyAccessDecoratorTest.php [hunk 10]
```diff
@@ -138,6 +172,28 @@ public function testCreateViewPreferredChoicesAsPropertyPath()
...
+    /**
+     * @group legacy
+     */
+    public function testCreateViewPreferredChoicesAsPropertyPathWithCallableString()
+    {
+        $list = $this->getMock('Symfony\Component\Form\ChoiceList\ChoiceListInterface');
+
+        $this->decoratedFactory->expects($this->once())
+            ->method('createView')
+            ->with($list, 'end')
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/ChoiceList/Factory/PropertyAccessDecoratorTest.php]: Your tests fail because it should be `->with($list, $this->isInstanceOf('\Closure'))`.

**Comment by webmozart** [comment]
@xabbuh I don't think we should add any extra support for string callables. It was never intended to support this functionality.

**Comment by Simperfit** [comment]
@HeahDude Thank you :).

I will changed this PR when the PR from @HeahDude gets merged.

**Comment by xabbuh** [comment]
@webmozart If it never was intended to work that way, why should we then bother to make it work now (I mean the places where we do not only trigger the deprecations but add extra code)?

**Comment by webmozart** [comment]
@xabbuh We shouldn't :) Triggering deprecations only is enough for me.

**Comment by HeahDude** [comment]
@Simperfit it turns out I explored a dead end. Can you please remove the closures and update your tests ?
Thank you and sorry for that.

**Comment by Simperfit** [comment]
NP @HeahDude, i've modified the PR with that

File: src/Symfony/Component/Form/Tests/ChoiceList/Factory/PropertyAccessDecoratorTest.php [hunk 11]
```diff
@@ -349,4 +503,5 @@ public function testCreateViewAttrAsPropertyPathInstance()
             new PropertyPath('property')
         ));
     }
+
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/ChoiceList/Factory/PropertyAccessDecoratorTest.php]: This extra line should be removed.

**Comment by Simperfit** [comment]
@webmozart Is it ok like this ?

File: src/Symfony/Component/Form/Tests/ChoiceList/Factory/PropertyAccessDecoratorTest.php (hunk 11 above)
Comment (Simperfit) [inline src/Symfony/Component/Form/Tests/ChoiceList/Factory/PropertyAccessDecoratorTest.php]: Thank you, modified :)

**Comment by xabbuh** [comment]
:+1: LGTM

**Comment by HeahDude** [comment]
@xabbuh we should maybe wait for advice from @webmozart concerning:

>I would like to stick with a deprecation error here (which can be added into the default factory, yes).

ref https://github.com/symfony/symfony/pull/18057#issuecomment-194214677.

This one only targets the `PropertyAccessDecorator` for now.

File: src/Symfony/Component/Form/Tests/ChoiceList/Factory/PropertyAccessDecoratorTest.php [hunk 12]
```diff
@@ -63,6 +63,21 @@ public function testCreateFromChoicesPropertyPathInstance()
...
+     * @group legacy
+     */
+    public function testCreateFromChoicesPropertyPathWithCallableString()
+    {
+        $choices = array(array('end' => 'value'));
+
+        $this->decoratedFactory->expects($this->once())
+            ->method('createListFromChoices')
+            ->with($choices, 'end')
+            ->will($this->returnValue(array('value')));
```
Comment (webmozart) [inline src/Symfony/Component/Form/Tests/ChoiceList/Factory/PropertyAccessDecoratorTest.php]: I think this test case is a bit confusing. In fact, the actal values of the choices and return values don't matter here. This test suggests they do, however.

I'd set `$choices` to some arbitrary value, to make that clear, e.g. `array('foo', 'bar')`.

I'd also set the return value of this line to `'RESULT'` to make it clear that the result of the underlying implementation is simply passed through.

Comment (webmozart) [inline src/Symfony/Component/Form/Tests/ChoiceList/Factory/PropertyAccessDecoratorTest.php]: btw you can use `willReturn('RESULT')` here.

File: src/Symfony/Component/Form/Tests/ChoiceList/Factory/PropertyAccessDecoratorTest.php [hunk 13]
```diff
@@ -77,6 +92,25 @@ public function testCreateFromLoaderPropertyPath()
...
+        $loader = $this->getMock('Symfony\Component\Form\ChoiceList\Loader\ChoiceLoaderInterface');
+
+        $this->decoratedFactory->expects($this->once())
+            ->method('createListFromLoader')
+            ->with($loader, 'end')
+            ->will($this->returnCallback(function ($loader, $callback) {
+                $array = array('property' => 'value');
+
+                return $callback($array);
+            }));
```
Comment (webmozart) [inline src/Symfony/Component/Form/Tests/ChoiceList/Factory/PropertyAccessDecoratorTest.php]: same here: `willReturn('RESULT')`

**Comment by Simperfit** [comment]
@webmozart I've modified the PR with your help, thanks for that, does it looks good to you ?

**Comment by webmozart** [comment]
:+1: 

Status: Reviewed

File: src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php [hunk 14]
```diff
@@ -86,6 +86,8 @@ public function createListFromChoices($choices, $value = null)
     {
         if (is_string($value) && !is_callable($value)) {
             $value = new PropertyPath($value);
+        } elseif (is_string($value) && is_callable($value)) {
+            @trigger_error('Passing callable strings is deprecated since version 3.1 and PropertyAccessDecorator will treat them as strings in 4.0. You should use a "\Closure" instead.', E_USER_DEPRECATED);
```
Comment (stof) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: it will treat them as property paths.

Comment (Simperfit) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: I've modified the message :)

**Comment by Tobion** [comment]
As I said in #17993 we do not need to deprecate that behavior but could also just change it as a bugfix because we are going back to how it originally was.

**Comment by Tobion** [comment]
And ChoiceListFactoryInterface should be clarified that only array and Closure callables are supported. But not strings.

**Comment by Simperfit** [comment]
@Tobion Should i modify anything ?

**Comment by fabpot** [comment]
@Tobion Can you check this PR before merging? Thanks.

File: src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php [hunk 15]
```diff
@@ -114,9 +116,11 @@ public function createListFromChoices($choices, $value = null)
      * @return ChoiceListInterface The choice list
      */
     public function createListFromLoader(ChoiceLoaderInterface $loader, $value = null)
-    {
+    {
```
Comment (fabpot) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: extra space here

**Comment by fabpot** [comment]
The other option is to revert @Tobion 's PR which was merged about a year ago in 2.7 (#15561).

I would vote for the revert. ping @symfony/deciders

**Comment by HeahDude** [comment]
I would say this is ambiguous both ways:

 - If we revert Tobias' patch, the `PropertyAccessor` might fail if the string is a callable (original issue).

 - if we merge this PR, a string 'range' might not be transformed to property path if it's a callable, but we can pass 'new PropertyPath('range'), so `$choice->getRange()` is called

On second thought, I'm :+1: for deprecations and better document property path as choice option. (ref https://github.com/symfony/symfony-docs/pull/6393)

**Comment by HeahDude** [comment]
If this one should be merged, I think it misses notes in changelogs and "update to 4.0".

**Comment by Simperfit** [comment]
@HeahDude should I modify the to guide to update with this PR ?

**Comment by HeahDude** [comment]
Probably let's just wait for a symfony decider to make that call :)

**Comment by HeahDude** [comment]
If I'm not wrong, as I've just learned with another PR, you should modify the `CHANGELOG` file of the Form component to just note the deprecation in 3.1 section:
`* Using callable strings as choice options in ChoiceType has been deprecated and will no longer be supported in 4.0`

And a note in the root `UPGRADE 4.0` to explain:
```rst
* Using callable strings as choice options in ChoiceType has been deprecated since 3.1 and is no longer supported. Use a "\Closure" instead.

    Before:

    'choice_label' => 'range',

    After:

    'choice_label' => function ($choice) {
        return $choice->getRange();
    },
```

**Comment by Simperfit** [comment]
Got it, thank you will do that @HeahDude  :+1:

**Comment by HeahDude** [comment]
@Simperfit I've edited my comment, please use this new version instead. Thanks :)

**Comment by Simperfit** [comment]
@HeahDude I've added them

File: src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php (hunk 15 above)
Comment (Simperfit) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: I've removed them

File: UPGRADE-4.0.md [hunk 16]
```diff
@@ -15,7 +15,22 @@ Form
  * The `choices_as_values` option of the `ChoiceType` has been removed.
  * Support for data objects that implements both `Traversable` and
    `ArrayAccess` in `ResizeFormListener::preSubmit` method has been removed
-
+ * Using callable strings as choice options in ChoiceType has been deprecated
```
Comment (HeahDude) [inline UPGRADE-4.0.md]: Missing empty line before `* Using...`

File: UPGRADE-4.0.md [hunk 17]
```diff
@@ -15,7 +15,22 @@ Form
...
+  in favor of `PropertyPath`. Use a "\Closure" instead
+  
+  Before:
```
Comment (HeahDude) [inline UPGRADE-4.0.md]: You need to indent with 6 spaces the following example

File: src/Symfony/Component/Form/CHANGELOG.md [hunk 18]
```diff
@@ -7,6 +7,8 @@ CHANGELOG
  * deprecated the "choices_as_values" option of ChoiceType
  * deprecated support for data objects that implements both `Traversable` and
    `ArrayAccess` in `ResizeFormListener::preSubmit` method
+ * Using callable strings as choice options in `ChoiceType` has been deprecated and
```
Comment (HeahDude) [inline src/Symfony/Component/Form/CHANGELOG.md]: Please break the line if a word gets the line over 80 characters

**Comment by HeahDude** [comment]
Minor comments, then you'll need to rebase on current master because there is now some conflicts.

Thanks

File: UPGRADE-4.0.md (hunk 17 above)
Comment (HeahDude) [inline UPGRADE-4.0.md]: and "Before" too

File: UPGRADE-4.0.md [hunk 19]
```diff
@@ -15,7 +15,22 @@ Form
  * The `choices_as_values` option of the `ChoiceType` has been removed.
  * Support for data objects that implements both `Traversable` and
    `ArrayAccess` in `ResizeFormListener::preSubmit` method has been removed
-
+ * Using callable strings as choice options in ChoiceType has been deprecated
+  in favor of `PropertyPath`. Use a "\Closure" instead
```
Comment (HeahDude) [inline UPGRADE-4.0.md]: you missed the colon :)

File: UPGRADE-4.0.md (hunk 17 above)
Comment (Simperfit) [inline UPGRADE-4.0.md]: I don't understand what you want to indent to 6 spaces

File: UPGRADE-4.0.md [hunk 20]
```diff
@@ -16,6 +16,23 @@ Form
...
+   
+ * Using callable strings as choice options in ChoiceType has been deprecated
+  in favor of `PropertyPath`. Use a "\Closure" instead
+ 
+   Before:
+   ```php
+  'choice_value' => new PropertyPath('range'),
+  'choice_label' => 'strtoupper',
+   ```
+  After:
```
Comment (HeahDude) [inline UPGRADE-4.0.md]: Sorry, should be 3, but still missing 1 here :)

Comment (Simperfit) [inline UPGRADE-4.0.md]: i've added one space.

File: UPGRADE-4.0.md [hunk 21]
```diff
@@ -16,6 +16,23 @@ Form
 
  * Support for data objects that implements both `Traversable` and
    `ArrayAccess` in `ResizeFormListener::preSubmit` method has been removed.
+   
+ * Using callable strings as choice options in ChoiceType has been deprecated
+  in favor of `PropertyPath`. Use a "\Closure" instead
+ 
+    Before:
```
Comment (HeahDude) [inline UPGRADE-4.0.md]: Ok now you got one extra here

File: UPGRADE-4.0.md [hunk 22]
```diff
@@ -16,6 +16,23 @@ Form
...
+   ```php
+   'choice_value' => new PropertyPath('range'),
+   'choice_label' => 'strtoupper',
+   ```
+   After:
+   ```php
+   'choice_value' => 'range',
+   'choice_label' => function ($choice) {
+   return strtoupper($choice);
```
Comment (HeahDude) [inline UPGRADE-4.0.md]: and still missing four here

**Comment by Simperfit** [comment]
I've rebased and addes/removed the space :p

File: UPGRADE-4.0.md (hunk 22 above)
Comment (HeahDude) [inline UPGRADE-4.0.md]: You need to get rid of this comment :)

Comment (Simperfit) [inline UPGRADE-4.0.md]: done :)

File: UPGRADE-4.0.md [hunk 23]
```diff
@@ -16,6 +16,26 @@ Form
...
+    'choice_value' => new PropertyPath('range'),
+    'choice_label' => 'strtoupper',
+    ```
+ 
+    After:
+   
+    ```php
+    'choice_value' => 'range',
+    'choice_label' => function ($choice) {
+    return strtoupper($choice);
```
Comment (HeahDude) [inline UPGRADE-4.0.md]: You misunderstood the "four" space here mean four for this line only.

Now on got one extra space from line 23 to 37 and missing 3 here one 35 :p

Sorry, after that we're done, thank you for working on this PR :)

Comment (Simperfit) [inline UPGRADE-4.0.md]: I hope this time it's ok :p

Np :)

File: UPGRADE-4.0.md [hunk 24]
```diff
@@ -55,6 +75,8 @@ FrameworkBundle
     - `"form.type.submit"`
     - `"form.type.reset"`
 
+   `ArrayAccess` in `ResizeFormListener::preSubmit` method has been removed
```
Comment (HeahDude) [inline UPGRADE-4.0.md]: Why did you add this line here ? it should be form section and it misses the `*`

Comment (HeahDude) [inline UPGRADE-4.0.md]: to not conflict with master it must be before your change

Comment (Simperfit) [inline UPGRADE-4.0.md]: oops sorry

**Comment by HeahDude** [comment]
Looks perfect now :) Thanks @Simperfit!

This one is ready to merge now.

Deprecate or revert 470b1400d631ff167bf82851a7975e2299da0856 ? ping __@symfony/decoders__

**Comment by Simperfit** [comment]
ping @Tobion @fabpot @dunglas

File: UPGRADE-4.0.md [hunk 25]
```diff
@@ -16,6 +16,26 @@ Form
 
  * Support for data objects that implements both `Traversable` and
    `ArrayAccess` in `ResizeFormListener::preSubmit` method has been removed.
+   
+ * Using callable strings as choice options in ChoiceType has been deprecated
+   in favor of `PropertyPath`. Use a "\Closure" instead
```
Comment (xabbuh) [inline UPGRADE-4.0.md]: The sentence as is must be added to the `UPGRADE-3.1.md` file. In this file it should read like "Using callable strings as choice options in ChoiceType has been removed in favor of passing `PropertyPath` instances."

Comment (HeahDude) [inline UPGRADE-4.0.md]: :+1:

Comment (Simperfit) [inline UPGRADE-4.0.md]: is it ok like this @HeahDude @xabbuh ?

File: UPGRADE-4.0.md [hunk 26]
```diff
@@ -16,6 +16,25 @@ Form
 
  * Support for data objects that implements both `Traversable` and
    `ArrayAccess` in `ResizeFormListener::preSubmit` method has been removed.
+   
+ * Using callable strings as choice options in ChoiceType has been removed in favor of passing PropertyPath instances.
+
```
Comment (HeahDude) [inline UPGRADE-4.0.md]: I would replace "has been removed" by "is not supported anymore".

But please wait for a symfony decider to tell you if it's ok. Thanks :)

Comment (xabbuh) [inline UPGRADE-4.0.md]: You are right. That sounds a bit better.

Comment (Simperfit) [inline UPGRADE-4.0.md]: done :)

File: src/Symfony/Component/Form/CHANGELOG.md [hunk 27]
```diff
@@ -7,6 +7,9 @@ CHANGELOG
  * deprecated the "choices_as_values" option of ChoiceType
  * deprecated support for data objects that implements both `Traversable` and
    `ArrayAccess` in `ResizeFormListener::preSubmit` method
+   
+ * Using callable strings as choice options in `ChoiceType` has been deprecated
+  and will be used as `PropertyPath` instead of callable in Symfony 4.0.
```
Comment (fabpot) [inline src/Symfony/Component/Form/CHANGELOG.md]: missing space at the beginning of the line

Comment (Simperfit) [inline src/Symfony/Component/Form/CHANGELOG.md]: @fabpot  space added.

**Comment by Simperfit** [comment]
i've rebased with master.
ping @fabpot @webmozart @Tobion

**Comment by fabpot** [comment]
IIRC, we were waiting for @Tobion's opinion about merging or reverting instead.

**Comment by Tobion** [comment]
I would revert but I'm ok with both ways.

**Comment by fabpot** [comment]
Thank you @Simperfit.

**Comment by Simperfit** [comment]
NP @fabpot.
